{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Application of matrix decomposition to biological data\n",
    "\n",
    "So far, we've used PCA and ICA on not truly biological datasets, now we'll try a real biological dataset by obtaining the data from a public database.\n",
    "\n",
    "## 1. Find the database and accession codes\n",
    "\n",
    "At the end of most recent papers, they'll put a section called \"**Accession Codes**\" or \"**Accession Numbers**\" which will list an identifying \n",
    "\n",
    "### Example data accession section from a Cell paper\n",
    "\n",
    "![Accession numbers in Cell journal](figures/accession_numbers_cell.png)\n",
    "\n",
    "### Example data accession section from a Nature Biotech paper\n",
    "![Accession codes in Nature Biotech journal](figures/accession_codes_buettner.png)\n",
    "\n",
    "Let's do this for the Shalek2013 paper. Note: For some \"older\" papers, the accession code may not be on the PDF version of the paper but on the online version only. What I usually do then is search for the title of the paper and go to the journal website.\n",
    "\n",
    "* What database was the data deposited to? \n",
    "* What is its' accession number?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Go to the data in the database\n",
    "\n",
    "If you search for the database and the accession number, the first result will usually be the database with the paper info and the deposited data! Below is an example search for \"Array Express E-MTAB-2805.\"\n",
    "\n",
    "![Example search for \"Array Express E-MTAB-2805\"](figures/buettner_search_accession.png)\n",
    "\n",
    "Search for its database and accession number and you should get to a page that looks like this:\n",
    "\n",
    "![GEO overview page for Shalek 2013](figures/shalek2013_geo.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Find the gene expression matrix\n",
    "\n",
    "Lately, for many papers, they *do* give a processed expression matrix in the accession database that you can use directly. Luckily for us, that's exactly what the authors of the Shalek 2013 dataset did. If you notice at the bottom of the page, there's a table of Supplementary files and one of them is called \"`GSE41265_allGenesTPM.txt.gz`\". The link below is the \"(ftp)\" link copied down with the command \"`wget`\" which I think of as short for \"web-get\" so you can download files from the internet with the command line.\n",
    "\n",
    "Run the cell below to download the text file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2016-06-04 11:38:46--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz\n",
      "           => 'GSE41265_allGenesTPM.txt.gz'\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 2607:f220:41e:250::12, 130.14.250.11\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|2607:f220:41e:250::12|:21... failed: Operation timed out.\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|130.14.250.11|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /geo/series/GSE41nnn/GSE41265/suppl ... done.\n",
      "==> SIZE GSE41265_allGenesTPM.txt.gz ... 1099290\n",
      "==> PASV ... done.    ==> RETR GSE41265_allGenesTPM.txt.gz ... done.\n",
      "Length: 1099290 (1.0M) (unauthoritative)\n",
      "\n",
      "GSE41265_allGenesTP 100%[=====================>]   1.05M   276KB/s   in 3.6s   \n",
      "\n",
      "2016-06-04 11:40:07 (294 KB/s) - 'GSE41265_allGenesTPM.txt.gz' saved [1099290]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use the unix command \"`ls`\" (short for \"listing\") to look around the files that are available here and prove to ourselves that we actually have the file we just downloaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.1_Introduction.ipynb\r\n",
      "2.2_Hierarchical_clustering.ipynb\r\n",
      "2.3_Matrix_Decomposition.ipynb\r\n",
      "2.4_Manifold_learning.ipynb\r\n",
      "2.5_Compare_unsupervised.ipynb\r\n",
      "2.6_Additional_reading.ipynb\r\n",
      "2.9_Application_of_matrix_decomposition_to_shalek2013.ipynb\r\n",
      "GSE41265_allGenesTPM.txt.gz\r\n",
      "\u001b[1m\u001b[36mfigures\u001b[m\u001b[m/\r\n",
      "\u001b[1m\u001b[36mpapers\u001b[m\u001b[m/\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See, \"`GSE41265_allGenesTPM.txt.gz`\" is there!\n",
    "\n",
    "Since the file ends in \"`.gz`\", this tells us its a \"gnu-zipped\" or \"gzipped\" (\"gee-zipped\") file, which is a specific flavor of \"zipping\" or compressing a file. We need to use a gnu-zipping-aware program to decompress the file, which is \"`gunzip`\" (\"gnu-unzip\").\n",
    "\n",
    "Run the next cell to unzip the file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "! gunzip GSE41265_allGenesTPM.txt.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's \"`ls`\" again to see what files have changed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.1_Introduction.ipynb\r\n",
      "2.2_Hierarchical_clustering.ipynb\r\n",
      "2.3_Matrix_Decomposition.ipynb\r\n",
      "2.4_Manifold_learning.ipynb\r\n",
      "2.5_Compare_unsupervised.ipynb\r\n",
      "2.6_Additional_reading.ipynb\r\n",
      "2.9_Application_of_matrix_decomposition_to_shalek2013.ipynb\r\n",
      "GSE41265_allGenesTPM.txt\r\n",
      "\u001b[1m\u001b[36mfigures\u001b[m\u001b[m/\r\n",
      "\u001b[1m\u001b[36mpapers\u001b[m\u001b[m/\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So now we have the unzipped version of the file, \"`GSE41265_allGenesTPM.txt`\". I wonder how much space they saved by zipping it?\n",
    "\n",
    "Let's use the flags \"`-l`\" for \"long listing\" which will show us the sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 7456\r\n",
      "-rw-r--r--   1 olga  staff     2084 May 24 16:04 2.1_Introduction.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   214633 Jun  3 12:09 2.2_Hierarchical_clustering.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   351682 Jun  3 12:16 2.3_Matrix_Decomposition.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   198471 May 24 19:29 2.4_Manifold_learning.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   163301 May 25 10:15 2.5_Compare_unsupervised.ipynb\r\n",
      "-rw-r--r--   1 olga  staff     1532 May 24 16:04 2.6_Additional_reading.ipynb\r\n",
      "-rw-r--r--   1 olga  staff     7137 Jun  4 12:16 2.9_Application_of_matrix_decomposition_to_shalek2013.ipynb\r\n",
      "-rw-r--r--   1 olga  staff  2866331 Jun  4 11:40 GSE41265_allGenesTPM.txt\r\n",
      "drwxr-xr-x  27 olga  staff      918 Jun  4 11:59 \u001b[1m\u001b[36mfigures\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   5 olga  staff      170 May 24 16:03 \u001b[1m\u001b[36mpapers\u001b[m\u001b[m/\r\n"
     ]
    }
   ],
   "source": [
    "ls -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "oof, this is in pure bytes and I can't convert to multiples of 1024 easily in my head (1024 bytes = 1 kilobyte, 1024 kilobytes = 1 megabtye, etc -  the 1000/byte is a lie that the hard drive companies use!). So let's use the `-h` flag, which tells the computer to do th conversion for us. We can combine multiple flags with the same dash, so\n",
    "\n",
    "    ls -l -h\n",
    "\n",
    "Can be shortened to:\n",
    "\n",
    "    ls -lh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 7464\r\n",
      "-rw-r--r--   1 olga  staff   2.0K May 24 16:04 2.1_Introduction.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   210K Jun  3 12:09 2.2_Hierarchical_clustering.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   343K Jun  3 12:16 2.3_Matrix_Decomposition.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   194K May 24 19:29 2.4_Manifold_learning.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   159K May 25 10:15 2.5_Compare_unsupervised.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   1.5K May 24 16:04 2.6_Additional_reading.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   8.6K Jun  4 12:18 2.9_Application_of_matrix_decomposition_to_shalek2013.ipynb\r\n",
      "-rw-r--r--   1 olga  staff   2.7M Jun  4 11:40 GSE41265_allGenesTPM.txt\r\n",
      "drwxr-xr-x  27 olga  staff   918B Jun  4 11:59 \u001b[1m\u001b[36mfigures\u001b[m\u001b[m/\r\n",
      "drwxr-xr-x   5 olga  staff   170B May 24 16:03 \u001b[1m\u001b[36mpapers\u001b[m\u001b[m/\r\n"
     ]
    }
   ],
   "source": [
    "ls -lh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, the file is 2.7 megabytes, and in the \"wget\" command we saw that the file was 1 megabyte, so the gzipping *did* save half the space! I bet that adds up over all the millions of files that GEO hosts.\n",
    "\n",
    "Anyways, let's get on with the analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Reading in the data file\n",
    "\n",
    "To read the gene expression matrix, we'll use \"`pandas`\" a Python package for \"Panel Data Analysis\" (as in panels of data), which is a fantastic library for working with dataframes, and is Python's answer to R's dataframes. We'll take this opportunity to import ALL of the python libaries that we'll use today.\n",
    "\n",
    "We'll read in the data using `pandas` and look at the first 5 rows of the dataframe with the dataframe-specific function `.head()`. Whenever I read a new table or modify a dataframe, I **ALWAYS** look at it to make sure it was correctly imported and read in, and I want you to get into the same habit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>S4</th>\n",
       "      <th>S5</th>\n",
       "      <th>S6</th>\n",
       "      <th>S7</th>\n",
       "      <th>S8</th>\n",
       "      <th>S9</th>\n",
       "      <th>S10</th>\n",
       "      <th>...</th>\n",
       "      <th>S12</th>\n",
       "      <th>S13</th>\n",
       "      <th>S14</th>\n",
       "      <th>S15</th>\n",
       "      <th>S16</th>\n",
       "      <th>S17</th>\n",
       "      <th>S18</th>\n",
       "      <th>P1</th>\n",
       "      <th>P2</th>\n",
       "      <th>P3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XKR4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019906</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AB338584</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B3GAT2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023441</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055452</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029448</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031654</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>42.150208</td>\n",
       "      <td>0.680327</td>\n",
       "      <td>0.022996</td>\n",
       "      <td>0.110236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NPL</th>\n",
       "      <td>72.008590</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>128.062012</td>\n",
       "      <td>0.095082</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>112.310234</td>\n",
       "      <td>104.329122</td>\n",
       "      <td>0.119230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.116802</td>\n",
       "      <td>0.104200</td>\n",
       "      <td>0.106188</td>\n",
       "      <td>0.229197</td>\n",
       "      <td>0.110582</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.109356</td>\n",
       "      <td>6.727028</td>\n",
       "      <td>14.525447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2</th>\n",
       "      <td>0.109249</td>\n",
       "      <td>0.172009</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.182703</td>\n",
       "      <td>0.076012</td>\n",
       "      <td>0.078698</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.093698</td>\n",
       "      <td>0.076583</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693459</td>\n",
       "      <td>0.010137</td>\n",
       "      <td>0.081936</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.086879</td>\n",
       "      <td>0.068174</td>\n",
       "      <td>0.062063</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 S1        S2          S3        S4        S5        S6  \\\n",
       "GENE                                                                      \n",
       "XKR4       0.000000  0.000000    0.000000  0.000000  0.000000  0.000000   \n",
       "AB338584   0.000000  0.000000    0.000000  0.000000  0.000000  0.000000   \n",
       "B3GAT2     0.000000  0.000000    0.023441  0.000000  0.000000  0.029378   \n",
       "NPL       72.008590  0.000000  128.062012  0.095082  0.000000  0.000000   \n",
       "T2         0.109249  0.172009    0.000000  0.000000  0.182703  0.076012   \n",
       "\n",
       "                  S7          S8        S9       S10    ...           S12  \\\n",
       "GENE                                                    ...                 \n",
       "XKR4        0.000000    0.000000  0.000000  0.000000    ...      0.000000   \n",
       "AB338584    0.000000    0.000000  0.000000  0.000000    ...      0.000000   \n",
       "B3GAT2      0.000000    0.055452  0.000000  0.029448    ...      0.000000   \n",
       "NPL       112.310234  104.329122  0.119230  0.000000    ...      0.000000   \n",
       "T2          0.078698    0.000000  0.093698  0.076583    ...      0.693459   \n",
       "\n",
       "               S13       S14       S15       S16       S17        S18  \\\n",
       "GENE                                                                    \n",
       "XKR4      0.000000  0.000000  0.000000  0.000000  0.000000   0.000000   \n",
       "AB338584  0.000000  0.000000  0.000000  0.000000  0.000000   0.000000   \n",
       "B3GAT2    0.000000  0.031654  0.000000  0.000000  0.000000  42.150208   \n",
       "NPL       0.116802  0.104200  0.106188  0.229197  0.110582   0.000000   \n",
       "T2        0.010137  0.081936  0.000000  0.000000  0.086879   0.068174   \n",
       "\n",
       "                P1        P2         P3  \n",
       "GENE                                     \n",
       "XKR4      0.000000  0.019906   0.000000  \n",
       "AB338584  0.000000  0.000000   0.000000  \n",
       "B3GAT2    0.680327  0.022996   0.110236  \n",
       "NPL       7.109356  6.727028  14.525447  \n",
       "T2        0.062063  0.000000   0.050605  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Alphabetical order is standard\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import pandas as pd\n",
    "from sklearn.decomposition import PCA, FastICA\n",
    "\n",
    "%matplotlib notebook\n",
    "\n",
    "shalek2013_expression = pd.read_table('GSE41265_allGenesTPM.txt', \n",
    "                                      index_col=0)  # index_col=0 sets the first column as the row names \n",
    "shalek2013_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we have 21 columns but pandas by default shows maximum of 20 so let's change the setting so we can see ALL of the samples instead of just skipping sample 11 (**S11**). We'll change this for rows, too, and why will become obvious in a second."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>S4</th>\n",
       "      <th>S5</th>\n",
       "      <th>S6</th>\n",
       "      <th>S7</th>\n",
       "      <th>S8</th>\n",
       "      <th>S9</th>\n",
       "      <th>S10</th>\n",
       "      <th>S11</th>\n",
       "      <th>S12</th>\n",
       "      <th>S13</th>\n",
       "      <th>S14</th>\n",
       "      <th>S15</th>\n",
       "      <th>S16</th>\n",
       "      <th>S17</th>\n",
       "      <th>S18</th>\n",
       "      <th>P1</th>\n",
       "      <th>P2</th>\n",
       "      <th>P3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XKR4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019906</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AB338584</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B3GAT2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023441</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.055452</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029448</td>\n",
       "      <td>0.024137</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.031654</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>42.150208</td>\n",
       "      <td>0.680327</td>\n",
       "      <td>0.022996</td>\n",
       "      <td>0.110236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NPL</th>\n",
       "      <td>72.008590</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>128.062012</td>\n",
       "      <td>0.095082</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>112.310234</td>\n",
       "      <td>104.329122</td>\n",
       "      <td>0.119230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.116802</td>\n",
       "      <td>0.104200</td>\n",
       "      <td>0.106188</td>\n",
       "      <td>0.229197</td>\n",
       "      <td>0.110582</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.109356</td>\n",
       "      <td>6.727028</td>\n",
       "      <td>14.525447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2</th>\n",
       "      <td>0.109249</td>\n",
       "      <td>0.172009</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.182703</td>\n",
       "      <td>0.076012</td>\n",
       "      <td>0.078698</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.093698</td>\n",
       "      <td>0.076583</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693459</td>\n",
       "      <td>0.010137</td>\n",
       "      <td>0.081936</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.086879</td>\n",
       "      <td>0.068174</td>\n",
       "      <td>0.062063</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 S1        S2          S3        S4        S5        S6  \\\n",
       "GENE                                                                      \n",
       "XKR4       0.000000  0.000000    0.000000  0.000000  0.000000  0.000000   \n",
       "AB338584   0.000000  0.000000    0.000000  0.000000  0.000000  0.000000   \n",
       "B3GAT2     0.000000  0.000000    0.023441  0.000000  0.000000  0.029378   \n",
       "NPL       72.008590  0.000000  128.062012  0.095082  0.000000  0.000000   \n",
       "T2         0.109249  0.172009    0.000000  0.000000  0.182703  0.076012   \n",
       "\n",
       "                  S7          S8        S9       S10       S11       S12  \\\n",
       "GENE                                                                       \n",
       "XKR4        0.000000    0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "AB338584    0.000000    0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "B3GAT2      0.000000    0.055452  0.000000  0.029448  0.024137  0.000000   \n",
       "NPL       112.310234  104.329122  0.119230  0.000000  0.000000  0.000000   \n",
       "T2          0.078698    0.000000  0.093698  0.076583  0.000000  0.693459   \n",
       "\n",
       "               S13       S14       S15       S16       S17        S18  \\\n",
       "GENE                                                                    \n",
       "XKR4      0.000000  0.000000  0.000000  0.000000  0.000000   0.000000   \n",
       "AB338584  0.000000  0.000000  0.000000  0.000000  0.000000   0.000000   \n",
       "B3GAT2    0.000000  0.031654  0.000000  0.000000  0.000000  42.150208   \n",
       "NPL       0.116802  0.104200  0.106188  0.229197  0.110582   0.000000   \n",
       "T2        0.010137  0.081936  0.000000  0.000000  0.086879   0.068174   \n",
       "\n",
       "                P1        P2         P3  \n",
       "GENE                                     \n",
       "XKR4      0.000000  0.019906   0.000000  \n",
       "AB338584  0.000000  0.000000   0.000000  \n",
       "B3GAT2    0.680327  0.022996   0.110236  \n",
       "NPL       7.109356  6.727028  14.525447  \n",
       "T2        0.062063  0.000000   0.050605  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.options.display.max_columns = 21\n",
    "pd.options.display.max_rows = 21\n",
    "shalek2013_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can see all the samples!\n",
    "\n",
    "Let's take a look at the full size of the matrix with `.shape`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27723, 21)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shalek2013_expression.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wow, ~28k rows! That must be the genes, while there are 18 single cell samples and 3 pooled samples as the columns.\n",
    "\n",
    "## 4. Verify that the matrix conforms to machine learning standards\n",
    "\n",
    "Okay so we have the genes as the rows and the samples as the columns. To make this compatible with machine learning algorithms, we need to transpose it so that the rows are the features (genes) and the columns are the samples (individual cells and bulk sequencing libraries). We'll do that with `.T`, and verify the shape, in addition to showing the top 5 rows.\n",
    "\n",
    "I like to both print the shape of the matrix in addition to showing the \"head\" so I can keep track of the number of columns or rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21, 27723)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GENE</th>\n",
       "      <th>XKR4</th>\n",
       "      <th>AB338584</th>\n",
       "      <th>B3GAT2</th>\n",
       "      <th>NPL</th>\n",
       "      <th>T2</th>\n",
       "      <th>T</th>\n",
       "      <th>PDE10A</th>\n",
       "      <th>1700010I14RIK</th>\n",
       "      <th>6530411M01RIK</th>\n",
       "      <th>PABPC6</th>\n",
       "      <th>...</th>\n",
       "      <th>AK085062</th>\n",
       "      <th>DHX9</th>\n",
       "      <th>RNASET2B</th>\n",
       "      <th>FGFR1OP</th>\n",
       "      <th>CCR6</th>\n",
       "      <th>BRP44L</th>\n",
       "      <th>AK014435</th>\n",
       "      <th>AK015714</th>\n",
       "      <th>SFT2D1</th>\n",
       "      <th>PRR18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>72.008590</td>\n",
       "      <td>0.109249</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.774638</td>\n",
       "      <td>23.520936</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>460.316773</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>39.442566</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.172009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.367391</td>\n",
       "      <td>1.887873</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>823.890290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.967412</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.023441</td>\n",
       "      <td>128.062012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.249858</td>\n",
       "      <td>0.313510</td>\n",
       "      <td>0.166772</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1002.354241</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.095082</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.354157</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.887003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1230.766795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.131215</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.182703</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.039263</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>131.077131</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1614.749122</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.242179</td>\n",
       "      <td>95.485743</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27723 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GENE  XKR4  AB338584    B3GAT2         NPL        T2    T  PDE10A  \\\n",
       "S1     0.0       0.0  0.000000   72.008590  0.109249  0.0     0.0   \n",
       "S2     0.0       0.0  0.000000    0.000000  0.172009  0.0     0.0   \n",
       "S3     0.0       0.0  0.023441  128.062012  0.000000  0.0     0.0   \n",
       "S4     0.0       0.0  0.000000    0.095082  0.000000  0.0     0.0   \n",
       "S5     0.0       0.0  0.000000    0.000000  0.182703  0.0     0.0   \n",
       "\n",
       "GENE  1700010I14RIK  6530411M01RIK  PABPC6  ...    AK085062      DHX9  \\\n",
       "S1              0.0            0.0     0.0  ...         0.0  0.774638   \n",
       "S2              0.0            0.0     0.0  ...         0.0  0.367391   \n",
       "S3              0.0            0.0     0.0  ...         0.0  0.249858   \n",
       "S4              0.0            0.0     0.0  ...         0.0  0.354157   \n",
       "S5              0.0            0.0     0.0  ...         0.0  0.039263   \n",
       "\n",
       "GENE   RNASET2B     FGFR1OP  CCR6       BRP44L  AK014435  AK015714     SFT2D1  \\\n",
       "S1    23.520936    0.000000   0.0   460.316773       0.0  0.000000  39.442566   \n",
       "S2     1.887873    0.000000   0.0   823.890290       0.0  0.000000   4.967412   \n",
       "S3     0.313510    0.166772   0.0  1002.354241       0.0  0.000000   0.000000   \n",
       "S4     0.000000    0.887003   0.0  1230.766795       0.0  0.000000   0.131215   \n",
       "S5     0.000000  131.077131   0.0  1614.749122       0.0  0.242179  95.485743   \n",
       "\n",
       "GENE  PRR18  \n",
       "S1      0.0  \n",
       "S2      0.0  \n",
       "S3      0.0  \n",
       "S4      0.0  \n",
       "S5      0.0  \n",
       "\n",
       "[5 rows x 27723 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shalek2013_expression = shalek2013_expression.T\n",
    "print(shalek2013_expression.shape)\n",
    "shalek2013_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Filter on bad genes and bad cells\n",
    "\n",
    "Okay, now we're ready to do some analysis! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "shalek2013_expression"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (single-cell-bioinformatics)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
