{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of RNA-Seq analysis\n",
    "\n",
    "## 1. Library quality control\n",
    "\n",
    "\n",
    "### `.fastq` file format\n",
    "\n",
    "#### PHRED64 quality scores\n",
    "\n",
    "\n",
    "* Input: Raw `.fastq` files\n",
    "* Program: FASTQC, SeQC\n",
    "* Output: Report (`.html`) and plots (`.png`) of library quality\n",
    "\n",
    "## 2. Adapter trimming\n",
    "\n",
    "* Input: Raw `.fastq` files\n",
    "* Program: `cutadapt`, `trimgalore`\n",
    "* Output: Adapter trimmed `.fastq` files\n",
    "\n",
    "## 3. Map to repetitive elements\n",
    "\n",
    "### `.sam` file format\n",
    "\n",
    "\"Sequence alignment map\"\n",
    "\n",
    "\n",
    "* Input: Trimmed `.fastq` files and repetitive element \"genome\"\n",
    "* Program: `STAR`\n",
    "* Output: Alignment `.sam` and unaligned `.fastq`\n",
    "\n",
    "Example:\n",
    "\n",
    "    \n",
    "\n",
    "## 4. Map to transcriptome\n",
    "\n",
    "* Input: `.fastq` of trimmed reads which didn't align to transcriptome\n",
    "* Program: `STAR`\n",
    "* Output: Alignment `.sam`\n",
    "\n",
    "Example:\n",
    "    \n",
    "    \n",
    "\n",
    "## 5. Create sorted, indexed `.bam` file\n",
    "\n",
    "\n",
    "### `.bam` file format\n",
    "\n",
    "All programs downstream of mapping require a sorted, indexed `.bam` file. You can also do these steps with your repetitive element-aligned `.sam` file if \n",
    "\n",
    "Think of creating a sorted, indexed bam file like making a dictionary. \n",
    "\n",
    "A raw aligned `.sam` file like a jumble of words in random order:\n",
    "\n",
    "    single\n",
    "    cell\n",
    "    bioinformatics\n",
    "    experiment\n",
    "    beyonce\n",
    "    sequencing\n",
    "    library\n",
    "    prep\n",
    "\n",
    "Sorting puts the words into alphabet order:\n",
    "\n",
    "    beyonce\n",
    "    bioinformatics\n",
    "    cell\n",
    "    experiment\n",
    "    library\n",
    "    prep\n",
    "    sequencing\n",
    "    single\n",
    "\n",
    "And indexing is like adding the alphabet tabs in a dictionary:\n",
    "\n",
    "    B:\n",
    "        beyonce\n",
    "        bioinformatics\n",
    "    C:\n",
    "        cell\n",
    "    E:\n",
    "        experiment\n",
    "    L: \n",
    "        library\n",
    "    P:\n",
    "        prep\n",
    "    S:\n",
    "        sequencing\n",
    "        single\n",
    "\n",
    "Instead of letters, indexing adds \"chromosome tabs\" so each program will know where, say chromosome 7 (`chr7`) starts and can jump to it right away.\n",
    "\n",
    "\n",
    "\n",
    "### 5.1 Sort `.sam`\n",
    "\n",
    "For deeply sequenced samples (>50 million reads), this will take quite some time\n",
    "\n",
    "* Input: `.sam` of transcriptome-aligned reads\n",
    "* Program: `samtools sort`\n",
    "* Output: Sorted `.bam`\n",
    "\n",
    "Example (`samtools v1.3.1`): \n",
    "\n",
    "    samtools sort sample01.sam > sample01.sorted.bam\n",
    "    \n",
    "### 5.2 Index `.sam`\n",
    "\n",
    "Indexing only works with sorted bam files\n",
    "\n",
    "* Input: `.bam` of *sorted*, transcriptome-aligned reads\n",
    "* Program: `samtools sort`\n",
    "* Output: Sorted `.bam`\n",
    "\n",
    "Example (`samtools v1.3.1`): \n",
    "\n",
    "    samtools sort sample01.sorted.bam\n",
    "\n",
    "This creates the file `sample01.sorted.bam.bai`\n",
    "    \n",
    "\n",
    "## 6. Quantify gene expression\n",
    "\n",
    "* Input: `.bam` of *sorted*, transcriptome-aligned reads\n",
    "* Program: `featureCounts`, Salmon, HTSeq, ...\n",
    "* Output: ?? some kind of table (comma- or tab-delimited file)\n",
    "\n",
    "## 7. Create gene expression matrix\n",
    "\n",
    "* Input: Individual cells' gene expression counts\n",
    "* Program: R, Python, `awk`, `cut`, `paste`\n",
    "* Output: A matrix of size (`n_samples` $\\times$ `n_features`)\n",
    "\n",
    "## 8. Cell-level QC\n",
    "\n",
    "* Input: Gene expression matrix\n",
    "* Program: R, Python, MATLAB\n",
    "* Output: Gene expression matrix with \"bad\" cells removed\n",
    "\n",
    "## 9. Cell-level QC\n",
    "\n",
    "* Input: Gene expression matrix\n",
    "* Program: R, Python, MATLAB\n",
    "* Output: Gene expression matrix with \"bad\" cells removed\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
