# Overview of RNA-Seq analysis

## 1. Library quality control

### `.fasta` file format

This is the simple, very commonly used file format in bioinformatics for storing sequences. An example of the p53 protein sequence is below.

    >sp|P04637|P53_HUMAN Cellular tumor antigen p53 OS=Homo sapiens GN=TP53 PE=1 SV=4
    MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGP
    DEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAK
    SVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHE
    RCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNS
    SCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELP
    PGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPG
    GSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD

### `.fastq` file format

The current standard of sequencing files is [`fastq`], which is derived from the `fasta` format. Check out an example `fastq` format below which shows 3 entries:

    $ zcat 250662/SRR578548_1.fastq.gz | head -n 12
    @SRR578548.1 C0G7VACXX120322:8:2306:19628:74804 length=101
    TAAATTTTAAGTAAATGTTTAAGGGATTTTACACCGGTCTATGGAGGTTTGCATGTGTAATTTTACCTCTAATTAATTATAAGGCCAGGACCAAACCTTTC
    +SRR578548.1 C0G7VACXX120322:8:2306:19628:74804 length=101
    @88DD?DDFAA<C4:C>C@IE<C@F3+CF9CHHB9CDF@F<DF<*0)?(0?<4BC=FCEFIIIC=A)=CEE>?CA7?EE@?;?B@BBC;;5=9=?5?BA>A
    @SRR578548.2 C0G7VACXX120322:8:2203:18331:38524 length=101
    CCTTAAATTTTAAGTAAATGTTTAAAGGATTTTACACCGGTCTATGGAGGTTTGCATGTGTAATTTTACCTCTAATTAATTATAAGGCCAGGACCAAACCT
    +SRR578548.2 C0G7VACXX120322:8:2203:18331:38524 length=101
    CCCFFFFFHHHHHJHIJJJJIJJIJJJJHJJJJJJJJJJJFHGIJJJJJIGIJJJJJJIIHIIIIJJJJJJJJHHHHHHHFFFFFFFEDDDDDDDDDDDDD
    @SRR578548.3 C0G7VACXX120322:8:2206:17405:142266 length=101
    CCTTAAATTTTAAGTAAATGTTTAAGGGATTTTACACCGGTCTATGGAGGTTTGCATGTGTAATTTTACCTCTAATTAATTATAAGGCCAGGACCAAACCT
    +SRR578548.3 C0G7VACXX120322:8:2206:17405:142266 length=101
    CCCDFFFFGHHFFIHHIGJGJIIJIJJJGIHIJJFJJJIGHHIIGGGHJJ?BHHIJJJHH=FGIIIJJIJJGHHHHHHGFFFFFDDEEDD?BDDBDDDDDC
    
Notice that each read takes up 4 lines, where

    @unique_identifier length=101
    [genomic sequence]
    +unique_identifier length=101
    [ASCII-encoded PHRED quality score]

Let's break the quality down a bit.

### Sequencing quality scores

The quality of base calling for each position in the read is encoded as a score $Q$ from 1 to 40. Defining $p$ as the probability of incorrect base calling, then

$
Q = -10 \log_{10}(p)
$

Given any score, we can back-calculate the probability of an incorrect base by using our understanding of logarithms.

$
40 &= -10 \log_{10}(p)
\frac{40}{-10} &= \frac{-10}{-10} \log_{10}(p)
-4 &= \log_{10}(p)
10{-4} &= 10^{\log_{10}(p)}
10^{-4} &= p
$

Thus the best score of 40 represents the lowest probability of incorrect base calling, $p=10^{-4}$.

#### Exercise 1

What is the probabilty of incorrect base calling for a score of 39? For 31?

### PHRED64 quality scores

Because "40" takes up two spaces, instead of using numbers the score is encoded using letters in a system called "ASCII." You already know that computers store information in zeros and ones, which can be turned into numbers, well, the letter "!" (exclamation point) represents the number 64 and acts as the lowest value for the quality scores, while 

    !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHI

* Input: Raw `.fastq` files
* Program: FASTQC, SeQC
* Output: Report (`.html`) and plots (`.png`) of library quality

## 2. Adapter trimming

* Input: Raw `.fastq` files
* Program: `cutadapt`, `trimgalore`
* Output: Adapter trimmed `.fastq` files

## 3. Map to repetitive elements

### `.sam` file format

"Sequence alignment map"


* Input: Trimmed `.fastq` files and repetitive element "genome"
* Program: `STAR`
* Output: Alignment `.sam` and unaligned `.fastq`

Example:

    

## 4. Map to transcriptome

* Input: `.fastq` of trimmed reads which didn't align to transcriptome
* Program: `STAR`
* Output: Alignment `.sam`

Example:
    
    

## 5. Create sorted, indexed `.bam` file


### `.bam` file format

All programs downstream of mapping require a sorted, indexed `.bam` file. You can also do these steps with your repetitive element-aligned `.sam` file if 

Think of creating a sorted, indexed bam file like making a dictionary. 

A raw aligned `.sam` file like a jumble of words in random order:

    single
    cell
    bioinformatics
    experiment
    beyonce
    sequencing
    library
    prep

Sorting puts the words into alphabet order:

    beyonce
    bioinformatics
    cell
    experiment
    library
    prep
    sequencing
    single

And indexing is like adding the alphabet tabs in a dictionary:

    B:
        beyonce
        bioinformatics
    C:
        cell
    E:
        experiment
    L: 
        library
    P:
        prep
    S:
        sequencing
        single

Instead of letters, indexing adds "chromosome tabs" so each program will know where, say chromosome 7 (`chr7`) starts and can jump to it right away.



### 5.1 Sort `.sam`

For deeply sequenced samples (>50 million reads), this will take quite some time

* Input: `.sam` of transcriptome-aligned reads
* Program: `samtools sort`
* Output: Sorted `.bam`

Example (`samtools v1.3.1`): 

    samtools sort sample01.sam > sample01.sorted.bam
    
### 5.2 Index `.sam`

Indexing only works with sorted bam files

* Input: `.bam` of *sorted*, transcriptome-aligned reads
* Program: `samtools sort`
* Output: Sorted `.bam`

Example (`samtools v1.3.1`): 

    samtools sort sample01.sorted.bam

This creates the file `sample01.sorted.bam.bai`
    

## 6. Quantify gene expression

### Units

TPM, FPKM, what? Check out this blog post on [What the FPKM?](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/) to learn about RNA-seq units in depth.

* Input: `.bam` of *sorted*, transcriptome-aligned reads
* Program: `featureCounts`, Salmon, HTSeq, ...
* Output: ?? some kind of table (comma- or tab-delimited file)

## 7. Create gene expression matrix

* Input: Individual cells' gene expression counts
* Program: R, Python, `awk`, `cut`, `paste`
* Output: A matrix of size (`n_samples` $\times$ `n_features`)

## 8. Cell-level QC

* Input: Gene expression matrix
* Program: R, Python, MATLAB
* Possible criteria: 
    * Low number of reads mapped (indicates lowly sequenced)
    * Low percentage of reads mapped (indicates poor library)
    * High mitochondrial:genome ratio (indicates dying cell)
    * High spike-in:genome ratio (indicates low cellular mRNA content)
    * Low numbers of expressed genes (indicates low detection)
* Output: Gene expression matrix with "bad" cells removed

## 9. Gene-level QC

* Input: Gene expression matrix with "bad" cells removed
* Program: R, Python, MATLAB
* Possible criteria:
    * Low expression (e.g. $< 1$ TPM in $> 5$ cells)
* Output: Gene expression matrix with "bad" genes removed


## 10. Exploratory analysis

* Input: Gene expression matrix with "bad" genes removed
* Program: R, Python, MATLAB
* Possible algorithms:
    * Dimensionality reduction
        * PCA
        * ICA
        * MDS
        * t-SNE
* Output: plots


## 11. Subpopulation identification

* Input: Gene expression matrix with "bad" genes removed
* Program: R, Python, MATLAB
* Possible algorithms:
    * Clustering
        * Hierarchical clustering
        * $K$-means
    * Pseudo-time
        * Monocle
        * Waterfall
        * Wanderlust
        * Wishbone
* Output: plots, cluster identification


## 12. Subpopulation characterization

* Input: Gene expression matrix with "bad" genes removed, cluster or psuedotime ordering
* Program: R, Python, MATLAB
* Possible algorithms:
    * Differential expression
        * DESeq
        * scDE ("single-cell differential expression")
    * Classifiers
        * Support Vector Machines (SVM)
        * DecisionTree classifiers
    * Clustering
        * Hierarchical clustering on ordered gene expression
* Output: Lists of genes

## 13. Interpretation

* Input: Lists of genes
* Program: Biologist, Gene Ontology (e.g. DAVID) or other gene set enrichment analysis, PubMed
* Output: Biological description of populations
