# Overview of RNA-Seq analysis

## 1. Library quality control


### `.fastq` file format

#### PHRED64 quality scores


* Input: Raw `.fastq` files
* Program: FASTQC, SeQC
* Output: Report (`.html`) and plots (`.png`) of library quality

## 2. Adapter trimming

* Input: Raw `.fastq` files
* Program: `cutadapt`, `trimgalore`
* Output: Adapter trimmed `.fastq` files

## 3. Map to repetitive elements

### `.sam` file format

"Sequence alignment map"


* Input: Trimmed `.fastq` files and repetitive element "genome"
* Program: `STAR`
* Output: Alignment `.sam` and unaligned `.fastq`

Example:

    

## 4. Map to transcriptome

* Input: `.fastq` of trimmed reads which didn't align to transcriptome
* Program: `STAR`
* Output: Alignment `.sam`

Example:
    
    

## 5. Create sorted, indexed `.bam` file


### `.bam` file format

All programs downstream of mapping require a sorted, indexed `.bam` file. You can also do these steps with your repetitive element-aligned `.sam` file if 

Think of creating a sorted, indexed bam file like making a dictionary. 

A raw aligned `.sam` file like a jumble of words in random order:

    single
    cell
    bioinformatics
    experiment
    beyonce
    sequencing
    library
    prep

Sorting puts the words into alphabet order:

    beyonce
    bioinformatics
    cell
    experiment
    library
    prep
    sequencing
    single

And indexing is like adding the alphabet tabs in a dictionary:

    B:
        beyonce
        bioinformatics
    C:
        cell
    E:
        experiment
    L: 
        library
    P:
        prep
    S:
        sequencing
        single

Instead of letters, indexing adds "chromosome tabs" so each program will know where, say chromosome 7 (`chr7`) starts and can jump to it right away.



### 5.1 Sort `.sam`

For deeply sequenced samples (>50 million reads), this will take quite some time

* Input: `.sam` of transcriptome-aligned reads
* Program: `samtools sort`
* Output: Sorted `.bam`

Example (`samtools v1.3.1`): 

    samtools sort sample01.sam > sample01.sorted.bam
    
### 5.2 Index `.sam`

Indexing only works with sorted bam files

* Input: `.bam` of *sorted*, transcriptome-aligned reads
* Program: `samtools sort`
* Output: Sorted `.bam`

Example (`samtools v1.3.1`): 

    samtools sort sample01.sorted.bam

This creates the file `sample01.sorted.bam.bai`
    

## 6. Quantify gene expression

### Units

TPM, FPKM, what? Check out this blog post on [What the FPKM?](https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/) to learn about RNA-seq units in depth.

* Input: `.bam` of *sorted*, transcriptome-aligned reads
* Program: `featureCounts`, Salmon, HTSeq, ...
* Output: ?? some kind of table (comma- or tab-delimited file)

## 7. Create gene expression matrix

* Input: Individual cells' gene expression counts
* Program: R, Python, `awk`, `cut`, `paste`
* Output: A matrix of size (`n_samples` $\times$ `n_features`)

## 8. Cell-level QC

* Input: Gene expression matrix
* Program: R, Python, MATLAB
* Possible criteria: 
    * Low number of reads mapped (indicates lowly sequenced)
    * Low percentage of reads mapped (indicates poor library)
    * High mitochondrial:genome ratio (indicates dying cell)
    * High spike-in:genome ratio (indicates low cellular mRNA content)
    * Low numbers of expressed genes (indicates low detection)
* Output: Gene expression matrix with "bad" cells removed

## 9. Gene-level QC

* Input: Gene expression matrix with "bad" cells removed
* Program: R, Python, MATLAB
* Possible criteria:
    * Low expression (e.g. $< 1$ TPM in $> 5$ cells)
* Output: Gene expression matrix with "bad" genes removed


## 10. Exploratory analysis

* Input: Gene expression matrix with "bad" genes removed
* Program: R, Python, MATLAB
* Possible algorithms:
    * Dimensionality reduction
        * PCA
        * ICA
        * MDS
        * t-SNE
* Output: plots


## 11. Subpopulation identification

* Input: Gene expression matrix with "bad" genes removed
* Program: R, Python, MATLAB
* Possible algorithms:
    * Clustering
        * Hierarchical clustering
        * $K$-means
    * Pseudo-time
        * Monocle
        * Waterfall
        * Wanderlust
        * Wishbone
* Output: plots, cluster identification


## 12. Subpopulation characterization

* Input: Gene expression matrix with "bad" genes removed, cluster or psuedotime ordering
* Program: R, Python, MATLAB
* Possible algorithms:
    * Differential expression
        * DESeq
        * scDE ("single-cell differential expression")
    * Classifiers
        * Support Vector Machines (SVM)
        * DecisionTree classifiers
    * Clustering
        * Hierarchical clustering on ordered gene expression
* Output: Lists of genes

## 13. Interpretation

* Input: Lists of genes
* Program: Biologist, Gene Ontology (e.g. DAVID) or other gene set enrichment analysis, PubMed
* Output: Biological description of populations
